codegen/smithy-go-codegen/src/main/java/software/amazon/smithy/go/codegen/auth/GetIdentityMiddlewareGenerator.java

```diff
@@ -0,0 +1,108 @@
+/*
+ * Copyright 2023 Amazon.com, Inc. or its affiliates. All Rights Reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License").
+ * You may not use this file except in compliance with the License.
+ * A copy of the License is located at
+ *
+ *  http://aws.amazon.com/apache2.0
+ *
+ * or in the "license" file accompanying this file. This file is distributed
+ * on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either
+ * express or implied. See the License for the specific language governing
+ * permissions and limitations under the License.
+ */
+
+package software.amazon.smithy.go.codegen.auth;
+
+import static software.amazon.smithy.go.codegen.GoStackStepMiddlewareGenerator.createFinalizeStepMiddleware;
+import static software.amazon.smithy.go.codegen.GoWriter.goTemplate;
+
+import software.amazon.smithy.go.codegen.GoStdlibTypes;
+import software.amazon.smithy.go.codegen.GoWriter;
+import software.amazon.smithy.go.codegen.MiddlewareIdentifier;
+import software.amazon.smithy.go.codegen.SmithyGoTypes;
+import software.amazon.smithy.go.codegen.integration.ProtocolGenerator;
+import software.amazon.smithy.utils.MapUtils;
+
+public class GetIdentityMiddlewareGenerator {
+    public static final String MIDDLEWARE_NAME = "getIdentityMiddleware";
+    public static final String MIDDLEWARE_ID = "GetIdentity";
+
+    private final ProtocolGenerator.GenerationContext context;
+
+    public GetIdentityMiddlewareGenerator(ProtocolGenerator.GenerationContext context) {
+        this.context = context;
+    }
+
+    public static GoWriter.Writable generateAddMiddleware() {
+        return goTemplate("""
+                err = stack.Finalize.Add(&$L{
```

Comment:
I'm not 100% sure about this, but I don't think this middleware is needed.

In `generateSelectScheme()` of the auth scheme middleware, identity resolvers are already invoked to see if it's valid or not for choosing an auth scheme, so I think you can just attach that resolved identity to the `resolvedAuthScheme`.

Reply:
We check that the resolver _exists_ there at that time, but we don't actually get the identity until this middleware (hence `getIdentityMiddleware`).

We could definitely save the resolver on the first call as you've said though it's really only a string lookup.